src/server/pkg/sync/sync.go

```diff
@@ -191,3 +203,151 @@ func PushObj(pachClient pachclient.APIClient, commit *pfs.Commit, objClient obj.
 	}
 	return eg.Wait()
 }
+
+// PullObj pulls data from an object store into a pfs commit.
+func PullObj(pachClient pachclient.APIClient, commit *pfs.Commit, objClient obj.Client, root string) error {
+	var eg errgroup.Group
+	sem := make(chan struct{}, semCap)
+	if err := objClient.Walk(root, func(name string) error {
+		sem <- struct{}{}
+		eg.Go(func() (retErr error) {
+			defer func() { <-sem }()
```

Comment:
Why is this reading from `sem` deferred, instead of done at the beginning? If we're using `sem` to throttle, could you end up in a situation where you call `eg.Go()` on N objects, and all N functions passed to `eg.Go()` are run (flooding the object store), and then all of the functions compete to exit, but they've already done all of their reading so they're just taking up memory?